### PR TITLE
feat: vrl functions

### PR DESCRIPTION
**Pull request by oasisk**

### CODE REVIEW

File: src/job/mod.rs [hunk 1]
```diff
@@ -53,20 +53,21 @@ pub async fn init() -> Result<(), anyhow::Error> {
...
     }
 
+    let res = db::get_instance().await;
+    let instance_id;
+    if res.as_ref().is_err() || res.as_ref().unwrap().is_none() {
+        instance_id = ider::generate();
+        let _ = db::set_instance(&instance_id).await;
+    } else {
+        instance_id = res.unwrap().unwrap();
+    }
```
Comment (vvv) [inline src/job/mod.rs:63]: 1.
```suggestion
    let instance_id = match db::get_instance().await {
        Ok(Some(instance)) => instance,
        Ok(None) | Err(_) => {
            let id = ider::generate();
            let _ = db::set_instance(&id).await;
            id
        }
    };
```

2. `db::get_instance` never returns `Ok(None)`. Can we change its return type to `Result<String, anyhow::Error>`?

https://github.com/zinclabs/zincobserve/blob/18152e4d16eb6cfccd03d9ab8f0f1b5c9749cbed/src/service/db/mod.rs#L95-L104

File: src/service/ingestion/mod.rs [hunk 2]
```diff
@@ -174,6 +213,70 @@ pub async fn send_ingest_notification(mut trigger: Trigger, alert: Alert) {
...
+            if trans.trans_type == 0 {
+                let func = load_lua_transform(lua, trans.function.clone());
+                if let Some(local_fn) = func {
+                    stream_lua_map.insert(func_key, local_fn.to_owned());
+                }
+            } else {
+                let program = compile_vrl_function(&trans.function);
+                if let Some(local_pg) = program {
+                    stream_vrl_map.insert(func_key, local_pg.to_owned());
+                }
```
Comment (vvv) [inline src/service/ingestion/mod.rs:246]:
```suggestion
                if let Some(program) = compile_vrl_function(&trans.function) {
                    stream_vrl_map.insert(func_key, program.to_owned());
                }
```

**Review by vvv — APPROVED** [review]

File: src/job/mod.rs (hunk 1 above)
Comment (oasisk) [reply src/job/mod.rs:63]: will take this up in next PR

File: Cargo.toml [hunk 3]
```diff
@@ -20,6 +20,7 @@ version = "0.1.0"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [features]
+#default = ["zo_functions"]
```
[the remark this answers is not in the document]
Comment (hengfeiyang) [reply Cargo.toml:23]: if we default enable this features, we need to disable default-features when we build on windows. but it looks okay.

File: Cargo.toml [hunk 4]
```diff
@@ -136,6 +137,11 @@ uuid = { version = "1.2.2", features = ["v4", "fast-rng", "macro-diagnostics"] }
 walkdir = "2"
 zstd = "0.12.1"
 
+vrl = {  git  = "https://github.com/vectordotdev/vrl",version = "0.1.0" }
+value = { package = "value", git  = "https://github.com/vectordotdev/vrl",version = "0.1.0"  }
+stdlib = { package = "vrl-stdlib",  git  = "https://github.com/vectordotdev/vrl",version = "0.1.0"  }
+
```
[the remark this answers is not in the document]
Comment (oasisk) [reply Cargo.toml:143]: we need all three of them , renamed value -> vrl-value , stdlib -> vrl-stdlib

File: src/common/vrl_utils.rs [hunk 5]
```diff
@@ -0,0 +1,29 @@
+use chrono::SecondsFormat;
+
+use super::json;
+
+pub fn vrl_value_to_json_value(value: value::Value) -> json::Value {
```
[the remark this answers is not in the document]
Comment (oasisk) [reply src/common/vrl_utils.rs:5]: done

File: Cargo.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (hengfeiyang) [reply Cargo.toml:143]: maybe we can fork it and make some change and then publish it to rust cates.io

**Review by hengfeiyang — APPROVED** [review]

**Comment by syepes** [comment]
@oasisk WOW this is a great feature! why better then lua :-)
Can't wait to see some docs on this

**Pull request merged**